Add Muxed address support for transactions

leofelix077 opened the pull request:
Closes #492 
Closes #525 

Also fixes normal Transactions, sending standard tokens to Muxed address, which was not previously working on the app (works on the extension)


* Descriptions of each video are part of the video title

https://github.com/user-attachments/assets/9880c5f2-3cfb-4568-983e-39fc712f2262


https://github.com/user-attachments/assets/a085aade-ce1e-4631-8cd4-386abc8fbcaf


https://github.com/user-attachments/assets/2c196906-be72-40d1-8785-6cf6645b1c91


https://github.com/user-attachments/assets/0694b2d1-785d-4c49-b005-545558029519


https://github.com/user-attachments/assets/3651f840-9e6b-4831-8adb-a98d7c2b25d8


https://github.com/user-attachments/assets/f54aa080-cf0d-424c-b0ee-510b387e7ca1


--------------

Screens for the error and warning messages:


<img width="349" height="772" alt="memo-info-settings" src="https://github.com/user-attachments/assets/27a08f4e-c74e-4ed2-9ed8-d5c4d7c772fb" />
<img width="340" height="789" alt="M-address-memo-disabled" src="https://github.com/user-attachments/assets/3e9c011b-b4ab-4d3a-9f84-c7c2d63210f4" />
<img width="337" height="772" alt="memo-unsupported-old-token-or-C-address" src="https://github.com/user-attachments/assets/90ba9fa0-0862-4d2e-b1cb-bc3aaba79abc" />
<img width="347" height="773" alt="memo-unsupported-banner-review" src="https://github.com/user-attachments/assets/a80f4f04-f43b-43b7-8e85-d90a1678116c" />
<img width="338" height="781" alt="review-muxed-address-not-supported" src="https://github.com/user-attachments/assets/8f29ee3f-a5fd-4f9e-916b-746ffc5994c1" />




#### PR structure

- [x] This PR does not mix refactoring changes with feature changes (break it down into smaller PRs if not).
- [x] This PR has reasonably narrow scope (break it down into smaller PRs if not).
- [x] This PR includes relevant before and after screenshots/videos highlighting these changes.
- [x] I took the time to review my own PR.

#### Testing

- [ ] These changes have been tested and confirmed to work as intended on Android.
- [x] These changes have been tested and confirmed to work as intended on iOS.
- [ ] ~These changes have been tested and confirmed to work as intended on small iOS screens.~
- [ ] ~These changes have been tested and confirmed to work as intended on small Android screens.~
- [x] I have tried to break these changes while extensively testing them.
- [x] This PR adds tests for the new functionality or fixes.

#### Release

- [x] This is not a breaking change.
- [x] This PR updates existing JSDocs when applicable.
- [x] This PR adds JSDocs to new functionalities.
- [ ] I've checked with the product team if we should add metrics to these changes.
- [ ] I've shared relevant before and after screenshots/videos highlighting these changes with the design team and they've approved the changes.

aristidesstaffieri commented on src/services/transactionService.ts:
```diff
@@ -349,92 +374,158 @@ export const buildPaymentTransaction = async (
-          ? selectedBalance.decimals
-          : DEFAULT_DECIMALS;
+      // Convert human-readable amount to base units
+      // Example: 1.33 tokens with 6 decimals = 1.33 * 10^6 = 1,330,000 base units
```
is there any reason to not have this function take the raw value instead of the display?
It seems like a logic leak to have a service level function deal with display values.

leofelix077 commented on src/services/transactionService.ts:
```diff
@@ -273,34 +291,41 @@ export const buildSorobanTransferOperation = (
+    // transactionBuilder is mutated in place
     transactionBuilder.addOperation(transaction);
+
+    return finalDestination;
```
as far as I've seen the transactionBuilder is a class/singleton that builds the transaction internally. It's called from outside, so I don't need to return it after mutating

leofelix077 commented:
@sdfcharles @Keeeeeeeks can you take a look at the screenshots/copies if they are fine for you and/or suggestions? I made some changes since the last time we discussed on the Slack thread

CassioMG commented on src/services/transactionService.ts:
```diff
@@ -249,21 +258,30 @@ interface IBuildSorobanTransferOperation {
  */
 export const buildSorobanTransferOperation = (
   params: IBuildSorobanTransferOperation,
-): TransactionBuilder => {
```
looks like buildSorobanTransferOperation is only being used locally here, so we could remove the `export` to make it easier to review/debug

CassioMG commented on src/services/transactionService.ts:
```diff
@@ -349,92 +374,158 @@ export const buildPaymentTransaction = async (
+      let contractId: string;
+      let token: SdkToken;
+
+      if (isCustomToken && selectedBalance.contractId) {
```
```suggestion
      if (isCustomToken) {
```

isCustomToken already implies selectedBalance.contractId exists, is lint complaining here?

leofelix077 replied:
not really. leftover from initial development

CassioMG commented on src/services/transactionService.ts:
```diff
@@ -560,19 +652,32 @@ export const buildSendCollectibleTransaction = async (
-      xdr.ScVal.scvU32(tokenId), // token_id
+      new Address(senderAddress).toScVal(),
+      new Address(finalDestination).toScVal(),
+      xdr.ScVal.scvU32(tokenId),
```
```suggestion
      new Address(senderAddress).toScVal(), // from
      new Address(finalDestination).toScVal(), // to
      xdr.ScVal.scvU32(tokenId), // token_id
```

CassioMG commented on src/services/transactionService.ts:
```diff
@@ -560,19 +652,32 @@ export const buildSendCollectibleTransaction = async (
 
-    if (transactionMemo) {
-      txBuilder.addMemo(Memo.text(transactionMemo));
-    }
```
just checking, are we removing this code here because we don't allow memo editing on Swap flow?

leofelix077 replied:
yes. leftover from the initial memo implementation, but it's unreachable code

CassioMG commented on src/ducks/transactionBuilder.ts:
```diff
@@ -135,30 +136,47 @@ export const useTransactionBuilderStore = create<TransactionBuilderState>(
-        } else {
-          logger.warn(
-            "TransactionBuilderStore",
-            "Recipient is not a contract, using standard transaction XDR.",
```
thanks for removing this warning, it was not very helpful and was polluting Sentry logs

CassioMG commented on src/ducks/transactionBuilder.ts:
```diff
@@ -135,30 +136,47 @@ export const useTransactionBuilderStore = create<TransactionBuilderState>(
 
-        // Only update store if this build request is still the latest one.
-        // This prevents race conditions where a slow async response from
-        // an older transaction overwrites state from a newer one.
```
let's keep this comment please

CassioMG commented on src/ducks/transactionBuilder.ts:
```diff
@@ -179,8 +197,6 @@ export const useTransactionBuilderStore = create<TransactionBuilderState>(
         );
 
-        // Only set error state if this build request is still current.
-        // Prevents stale error from overwriting newer transaction state.
```
let's keep this comment please

CassioMG commented on src/ducks/transactionBuilder.ts:
```diff
@@ -307,11 +327,6 @@ export const useTransactionBuilderStore = create<TransactionBuilderState>(
-          "TransactionBuilderStore",
-          "Failed to build send collectible transaction",
-          error,
-        );
```
this one may be worth keeping, wdyt?

leofelix077 commented:
Android smoke test

https://github.com/user-attachments/assets/09b68ea8-297e-411e-9a3d-5973d797585e

CassioMG commented on src/components/screens/SendScreen/components/SendReviewBottomSheet.tsx:
```diff
@@ -48,9 +48,6 @@ type SendReviewBottomSheetProps = {
   onBannerPress?: () => void;
-  isMalicious?: boolean;
-  isSuspicious?: boolean;
-  isUnableToScan?: boolean;
```
is this removal intended?

CassioMG reviewed (approved):
Great work on this PR, lots of important fixes/improvements done 👏
